### ISSUE TITLE
Allow comments in @FILE files

### ISSUE THREAD

**Issue by somatronicon** [nunit/nunit#2080]
In the @FILE files, allow comments. These files could get pretty complex, and having the ability to add comments would be beneficial. Maybe if a line starts with # it's interpreted as a comment? Not sure, just throwing out ideas.

```
# Black list specific tests we don't want to run
--where "method != /.*_(Blah|Blah2|Blah3)$/"
```

Great project by the way :+1:

**Comment by JustinRChou** [comment]
@CharliePoole 
Is this really a easy fix and wouldn't this involve updating the console runner and the lite runner?

**Comment by CharliePoole** [comment]
I was thinking of this as a console issue but you're right that it applies to NUnitLite as well.

In the console, it's extremely easy to just ignore a full line starting with # - looks like a change of one line. Doing more is possible, but difficult since the individual line can contain # (or any other characters) in quotes. The format was designed to be simple and we probably shouldn't complicate it, since that would require a redesign without a clear benefit.

The NUnitLite implementation is entirely different and treats the entire file as a single string. That looks like an easyfix as well, but more than a single line.

I'll duplicate this under the console as well.

**Comment by somatronicon** [comment]
Thanks @CharliePoole and @JustinRChou!

**Comment by JustinRChou** [comment]
I am going to work on this.

And I am having trouble finding tests for the `@File` option?

**Comment by CharliePoole** [comment]
https://github.com/nunit/nunit/blob/master/src/NUnitFramework/nunitlite.tests/CommandLineTests.cs#L44

**Comment by JustinRChou** [comment]
I just saw the update for the console one nunit/nunit-console#202
Probably going to do a compare and move your changes over.

**Comment by JustinRChou** [comment]
@CharliePoole I tried moving your testcases over to `AtsignFilenameTests` but the existing testcases are already failing when I am running them in the Test Explorer but `build -t test` looks fine. 

I don't know what is going on here.

**Comment by CharliePoole** [comment]
@JustinRChou Have you run nunit tests under the TestExplorer before? I would be very surprised if that worked at all. When testing NUnit, we always tell people to stick with just the basics, no "third-party" stuff - which is kind of what the VS adapter is, although it belongs to us.

As far as the @file implementation, it is of course entirely different in the two programs, developed by two different people.

**Comment by JustinRChou** [comment]
Oh, that really explains a lot when I was working on the framework test cases.

**Comment by JustinRChou** [comment]
Looks like 
```
        internal static IEnumerable<string> GetArgs(string commandLine)
        {
            const string re = @"\G(""((""""|[^""])+)""|(\S+)) *";
            var ms = Regex.Matches(commandLine, re);
            return ms.Cast<Match>().Select(m => Regex.Replace(m.Groups[2].Success ? m.Groups[2].Value : m.Groups[4].Value, @"""""", @""""));
        }
```
is having issues dealing with spaces in `#This is a comment`, it does a split on the spaces so the result becomes`is a comment` .

**Comment by CharliePoole** [comment]
You may have to reimplement it in line with the console runner.

**Comment by CharliePoole** [comment]
@JustinRChou Are you still working on this?

**Comment by JustinRChou** [comment]
Yes.
I am still working on a test when there is an `@File` with a file. 
` [TestCase("--arg1 @file1.txt --arg2 @file2.txt", "file1.txt:--filearg1 --filearg2,file2.txt:--filearg3 @file3.txt,file3.txt:--filearg4", "--arg1 --filearg1 --filearg2 --arg2 --filearg3 --filearg4", "")]`
I ran that test on the console and it was failing too, I am not sure if it is due to test generation of the files or the parsing.

**Comment by CharliePoole** [comment]
Looking at the code, `@[,]` appears to be setting `,` as the delimiter for the file. It's not a supported feature in the console, which has a much simpler implementation. Personally, I prefer simple. 😄

**Comment by JustinRChou** [comment]
Well then, #1877 explains a lot.
The delimiter `[,]` part is removed in my branch.
So the nesting refers to the `@File` within a `@File`?
And you're keeping the nesting level to 3 or changing it to 1?

**Comment by CharliePoole** [comment]
Bear in mind I didn't write it - I can only look at the code. I don't recall that the console has any levels at all but I can't see the code right now.

**Comment by CharliePoole** [comment]
Maybe one of us should do #1877 first?

**Comment by JustinRChou** [comment]
On #1877, looks like only the nesting needs to be added to the console.
Let see if I can get the old nesting to work with what I copied from the console.

**Comment by CharliePoole** [comment]
Or remove nesting from nunitlite. Whoever takes it needs to lead a discussion whereby we decide.
 
The other big difference is that the console is line oriented whereas nunitlite is token-oriented.

**Comment by oznetmaster** [comment]
As I pointed out at the time, both myself and many of my users use nesting a significant amount of the time.

Since it pretty trivial to implement, as my nunitlite implementation shows, I think it should be added to the console

**Comment by CharliePoole** [comment]
@oznetmaster @JustinRChou I assigned #1877 to myself, at least for design purposes. I'll comment further there.

**Comment by CharliePoole** [comment]
@JustinRChou It looks like I inadvertently took care of this in resolving #1877, forgetting it was a separate issue. 😞 Please peek at it and make sure you agree.

**Comment by JustinRChou** [comment]
@CharliePoole 

Everything is fine.

**Comment by CharliePoole** [comment]
Fixed by #2130

**Comment by somatronicon** [comment]
Thanks again for getting this one pushed through all; nice job running a nice project :+1:
